From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Robert Vasek <rvasek01@gmail.com>
Date: Fri, 13 Mar 2026 13:51:17 +0100
Subject: [PATCH] UPSTREAM: <squash> fix calculating kcp storage annotations

Can be squashed with 'UPSTREAM: <carry>: storage/etcd3: clusterize'

On-behalf-of: SAP robert.vasek@sap.com
Signed-off-by: Robert Vasek <robert.vasek@clyso.com>
---
 .../apiserver/pkg/storage/etcd3/store.go      |   2 +-
 .../apiserver/pkg/storage/etcd3/store_kcp.go  | 121 +++++++++--
 .../pkg/storage/etcd3/store_kcp_test.go       | 202 ++++++++++++++++--
 .../apiserver/pkg/storage/etcd3/watcher.go    |   4 +-
 4 files changed, 287 insertions(+), 42 deletions(-)

diff --git a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
@@ -879,7 +879,7 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 
 			// kcp
 			clusterName := adjustClusterNameIfWildcard(shard, cluster, crdIndicator, keyPrefix, string(kv.Key))
-			shardName := adjustShardNameIfWildcard(shard, keyPrefix, string(kv.Key))
+			shardName := adjustShardNameIfWildcard(shard, cluster, crdIndicator, keyPrefix, string(kv.Key))
 			obj, err := s.decodeListItem(ctx, data, kv.ModRevision, s.codec, s.versioner, newItemFunc, clusterName, shardName)
 			if err != nil {
 				recordDecodeError(s.groupResource, string(kv.Key))
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp.go b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp.go
--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp.go
@@ -26,44 +26,124 @@ import (
 	"k8s.io/klog/v2"
 )
 
+// adjustKeyForPrefix returns storage path key stripped of its prefix, with corrections that
+// account for wildcards and partial metadata if they are part of the request. We need that
+// so the resulting key has a stable form, from which the shard and cluster name can be parsed.
+//
+// The key is in the following format:
+//
+//	<Storage prefix> / <Group> / <Resource> / [ <Identity or "customresources"> / ] [ <Shard> / ] <Cluster> / [ <Namespace> / ] <Name>
+//
+// The prefix is variable, and may include segments ranging from:
+//
+//	<Storage prefix> / <Group> / <Resource> /
+//
+// And up to:
+//
+//	<Storage prefix> / <Group> / <Resource> / [ <Identity or "customresources"> / ] [ <Shard> / ] <Cluster> /
+//
+// The exact prefix length depends on how inclusive the request is:
+//
+//   - Partial metadata requests accept any key with matching group-resource, regardless of its identity or
+//     CustomResource origin. This makes the prefix stop right after .../<Resource>/. Only applies if CRDRequest=true.
+//   - Shard wildcard requests accept any key with matching fully-qualified resource,
+//     which makes the prefix stop right after .../<Identity or "customresources">/ in case of CRDRequest=true,
+//     or right after .../<Resource>/ in case of CRDRequest=false.
+//   - Cluster wildcard requests accept any key with matching fully-qualified resource, which makes the prefix
+//     stop right .../<Shard>/, or .../<Identity or "customresources">/, or .../<Resource>/, depending on which
+//     of those is part of the request.
+//
+// This function checks all these cases, and returns the key in format:
+//
+//   - If the shard in request is a wildcard: <Shard> / <Cluster> / [ <Namespace> / ] <Name>
+//   - If the shard in request is empty or concrete:    <Cluster> / [ <Namespace> / ] <Name>
+func adjustKeyForPrefix(prefix, key string, crdRequest bool, cluster *genericapirequest.Cluster, shard genericapirequest.Shard) string {
+	popSegment := func(s string) string {
+		segmentStart := strings.IndexByte(s, '/')
+		if segmentStart < 0 {
+			return s
+		}
+		if segmentStart == len(s) {
+			return s
+		}
+		return s[segmentStart+1:]
+	}
+
+	keyWithoutPrefix := strings.TrimPrefix(key, prefix)
+
+	// This is what keyWithoutPrefix can now contain, if:
+	//
+	//   [ <Identity or "customresources"> / ] [ <Shard> / ] <Cluster> / [ <Namespace> / ] <Name>
+	//   ^                                     ^             ^           ^----------------------^
+	//   |                                     |             |                    |             |
+	//   |                                     |             |          (4) !cluster.Wildcard   |
+	//   |                                     |             +----------------------------------+
+	//   |                                     |                   |                            |
+	//   |                                     |            (3) shard.Empty()                   |
+	//   |                                     +------------------------------------------------+
+	//   |                                                 |                                    |
+	//   |                                        (2) !shard.Empty()                            |
+	//   +--------------------------------------------------------------------------------------+
+	//                  |
+	//          (1) cluster.PartialMetadataRequest (assumes crdRequest)
+
+	if !crdRequest || !cluster.PartialMetadataRequest {
+		// The prefix already includes <Identity or "customresources"> segment,
+		// or this is not a CRD request, so keyWithoutPrefix is clean.
+		return keyWithoutPrefix
+	}
+
+	// With partial metadata, the prefix ends BEFORE the identity segment.
+	keyWithoutIdentity := popSegment(keyWithoutPrefix)
+
+	if shard.Empty() || shard.Wildcard() {
+		// Shard needs to be part of the key if it is wildcard.
+		return keyWithoutIdentity
+	}
+
+	// We are scoped to a shard, but the <Shard> segment wasn't part of the prefix
+	// because of partial metadata ending it early, so we need to strip it from the key now.
+	keyWithoutShard := popSegment(keyWithoutIdentity)
+
+	return keyWithoutShard
+}
+
 // adjustClusterNameIfWildcard determines the logical cluster name. If this is not a cluster-wildcard list/watch request,
-// the cluster name is returned unmodified. Otherwise, the cluster name is extracted from the key based on whether it is
-// - a shard-wildcard request: <prefix>/shardName/clusterName/<remainder>
-// - CR partial metadata request: <prefix>/identity/clusterName/<remainder>
-// - any other request: <prefix>/clusterName/<remainder>.
+// the cluster name is returned unmodified. Otherwise, the cluster name is extracted from the storage key.
 func adjustClusterNameIfWildcard(shard genericapirequest.Shard, cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
 	if !cluster.Wildcard {
 		return cluster.Name
 	}
 
-	keyWithoutPrefix := strings.TrimPrefix(key, keyPrefix)
+	keyWithoutPrefix := adjustKeyForPrefix(keyPrefix, key, crdRequest, cluster, shard)
 	parts := strings.SplitN(keyWithoutPrefix, "/", 3)
 
 	extract := func(minLen, i int) logicalcluster.Name {
 		if len(parts) < minLen {
-			klog.Warningf("shard=%s cluster=%s invalid key=%s had %d parts, not %d", shard, cluster, keyWithoutPrefix, len(parts), minLen)
+			klog.Warningf("shard=%s cluster=%v invalid key=%s had %d parts, wanted %d", shard, cluster, keyWithoutPrefix, len(parts), minLen)
 			return ""
 		}
 		return logicalcluster.Name(parts[i])
 	}
 
-	switch {
-	case cluster.PartialMetadataRequest && crdRequest:
-		// expecting 2699f4d273d342adccdc8a32663408226ecf66de7d191113ed3d4dc9bccec2f2/root:org:ws/<remainder>
-		// OR customresources/root:org:ws/<remainder>
-		return extract(3, 1)
-	case shard.Wildcard():
-		// expecting shardName/clusterName/<remainder>
-		return extract(3, 1)
-	default:
-		// expecting root:org:ws/<remainder>
+	if !shard.Wildcard() {
+		// Shard is either empty or concrete. In both cases, adjustKeyForPrefix
+		// has already stripped it (along with <Identity> segment, if applicable).
+		//
+		// The remaining key is in format:
+		//   <Cluster> / <Remainder...>
 		return extract(2, 0)
 	}
+	// Shard is wildcard, and still present in the key.
+	//
+	// The remaining key is in format:
+	//   <Shard> / <Cluster> / <Remainder...>
+	return extract(3, 1)
 }
 
 // adjustShardNameIfWildcard determines a shard name. If this is not a shard-wildcard request,
 // the shard name is returned unmodified. Otherwise, the shard name is extracted from the storage key.
-func adjustShardNameIfWildcard(shard genericapirequest.Shard, keyPrefix, key string) genericapirequest.Shard {
+func adjustShardNameIfWildcard(shard genericapirequest.Shard, cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) genericapirequest.Shard {
 	if !shard.Empty() && !shard.Wildcard() {
 		return shard
 	}
@@ -74,10 +154,13 @@ func adjustShardNameIfWildcard(shard genericapirequest.Shard, keyPrefix, key str
 		return ""
 	}
 
-	keyWithoutPrefix := strings.TrimPrefix(key, keyPrefix)
+	keyWithoutPrefix := adjustKeyForPrefix(keyPrefix, key, crdRequest, cluster, shard)
+
+	// The remaining key is in format:
+	//   <Shard> / <Cluster> / <Remainder...>
 	parts := strings.SplitN(keyWithoutPrefix, "/", 3)
 	if len(parts) < 3 {
-		klog.Warningf("unable to extract a shard name, invalid key=%s had %d parts, not %d", keyWithoutPrefix, len(parts), 3)
+		klog.Warningf("unable to extract a shard name, invalid key=%s had %d parts, wanted %d", keyWithoutPrefix, len(parts), 3)
 		return ""
 	}
 	return genericapirequest.Shard(parts[0])
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp_test.go b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp_test.go
--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_kcp_test.go
@@ -53,12 +53,12 @@ func TestAdjustClusterNameIfWildcard(t *testing.T) {
 	for name, tc := range tests {
 		t.Run(name, func(t *testing.T) {
 			cluster := &genericapirequest.Cluster{
-				Name:                   logicalcluster.New("root:org:ws"),
+				Name:                   logicalcluster.Name("root:org:ws"),
 				PartialMetadataRequest: tc.partialMetadata,
 			}
 
 			if tc.wildcard {
-				cluster.Name = logicalcluster.Wildcard
+				cluster.Wildcard = true
 			}
 
 			key := "/registry/group/resource/identity/root:org:ws/somename"
@@ -79,53 +79,215 @@ func TestAdjustClusterNameIfWildcardWithShardSupport(t *testing.T) {
 	tests := map[string]struct {
 		cluster             genericapirequest.Cluster
 		shard               genericapirequest.Shard
+		crdRequest          bool
 		key                 string
 		keyPrefix           string
 		expectedClusterName string
 	}{
-		"not wildcard": {
-			cluster:             genericapirequest.Cluster{Name: logicalcluster.New("root:org:ws")},
+
+		// Built-in types.
+
+		"not wildcard, built-in type": {
+			cluster:             genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
 			shard:               "amber",
-			key:                 "/registry/group/resource:identity/amber/root:org:ws/somename",
-			keyPrefix:           "/registry/group/resource:identity/",
+			key:                 "/registry/core/configmaps/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/core/configmaps/amber/root:org:ws/",
 			expectedClusterName: "root:org:ws",
 		},
-		"both wildcard": {
-			cluster:             genericapirequest.Cluster{Name: logicalcluster.Wildcard},
+		"both wildcard, built-in type": {
+			cluster:             genericapirequest.Cluster{Wildcard: true},
 			shard:               "*",
-			key:                 "/registry/group/resource:identity/amber/root:org:ws/somename",
-			keyPrefix:           "/registry/group/resource:identity/",
+			key:                 "/registry/core/configmaps/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/core/configmaps/",
 			expectedClusterName: "root:org:ws",
 		},
-		"both wildcard, built-in type": {
-			cluster:             genericapirequest.Cluster{Name: logicalcluster.Wildcard},
+		"both wildcard, built-in type, partial": {
+			cluster:             genericapirequest.Cluster{Wildcard: true, PartialMetadataRequest: true},
 			shard:               "*",
 			key:                 "/registry/core/configmaps/amber/root:org:ws/somename",
 			keyPrefix:           "/registry/core/configmaps/",
 			expectedClusterName: "root:org:ws",
 		},
-		"only cluster wildcard": {
-			cluster:             genericapirequest.Cluster{Name: logicalcluster.Wildcard},
+		"only cluster wildcard, built-in type": {
+			cluster:             genericapirequest.Cluster{Wildcard: true},
 			shard:               "amber",
-			key:                 "/registry/group/resource:identity/amber/root:org:ws/somename",
-			keyPrefix:           "/registry/group/resource:identity/amber/",
+			key:                 "/registry/core/configmaps/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/core/configmaps/amber/",
 			expectedClusterName: "root:org:ws",
 		},
-		"only shard wildcard": {
-			cluster:             genericapirequest.Cluster{Name: logicalcluster.New("root:org:ws")},
+		"only cluster wildcard, built-in type, partial": {
+			cluster:             genericapirequest.Cluster{Wildcard: true, PartialMetadataRequest: true},
+			shard:               "amber",
+			key:                 "/registry/core/configmaps/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/core/configmaps/amber/",
+			expectedClusterName: "root:org:ws",
+		},
+		"only shard wildcard, built-in type": {
+			cluster:             genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
 			shard:               "*",
 			key:                 "/registry/core/configmaps/amber/root:org:ws/somename",
-			keyPrefix:           "/registry/group/resource:identity/",
+			keyPrefix:           "/registry/core/configmaps",
+			expectedClusterName: "root:org:ws",
+		},
+
+		// CRD types.
+
+		"both wildcard, CRD": {
+			cluster:             genericapirequest.Cluster{Wildcard: true},
+			shard:               "*",
+			crdRequest:          true,
+			key:                 "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/group/resource/identity/",
+			expectedClusterName: "root:org:ws",
+		},
+		"both wildcard, CRD, partial metadata": {
+			cluster:             genericapirequest.Cluster{Wildcard: true, PartialMetadataRequest: true},
+			shard:               "*",
+			crdRequest:          true,
+			key:                 "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/group/resource/",
+			expectedClusterName: "root:org:ws",
+		},
+
+		"only cluster wildcard, CRD": {
+			cluster:             genericapirequest.Cluster{Wildcard: true},
+			shard:               "amber",
+			crdRequest:          true,
+			key:                 "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/group/resource/identity/amber/",
+			expectedClusterName: "root:org:ws",
+		},
+		"only cluster wildcard, CRD, partial metadata": {
+			cluster:             genericapirequest.Cluster{Wildcard: true, PartialMetadataRequest: true},
+			shard:               "amber",
+			crdRequest:          true,
+			key:                 "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/group/resource/",
+			expectedClusterName: "root:org:ws",
+		},
+
+		"only shard wildcard, CRD": {
+			cluster:             genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
+			shard:               "*",
+			crdRequest:          true,
+			key:                 "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:           "/registry/group/resource/identity/",
 			expectedClusterName: "root:org:ws",
 		},
 	}
 
 	for name, tc := range tests {
 		t.Run(name, func(t *testing.T) {
-			clusterName := adjustClusterNameIfWildcard(tc.shard, &tc.cluster, false, tc.keyPrefix, tc.key)
+			clusterName := adjustClusterNameIfWildcard(tc.shard, &tc.cluster, tc.crdRequest, tc.keyPrefix, tc.key)
 			if tc.expectedClusterName != clusterName.String() {
 				t.Errorf("expected: %q, actual %q", tc.expectedClusterName, clusterName)
 			}
 		})
 	}
 }
+
+func TestAdjustShardNameIfWildcard(t *testing.T) {
+	tests := map[string]struct {
+		cluster       genericapirequest.Cluster
+		shard         genericapirequest.Shard
+		crdRequest    bool
+		key           string
+		keyPrefix     string
+		expectedShard genericapirequest.Shard
+	}{
+
+		// Built-in types.
+
+		"not wildcard, built-in type": {
+			cluster:       genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
+			shard:         "amber",
+			key:           "/registry/group/resource/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/amber/root:org:ws/",
+			expectedShard: "amber",
+		},
+		"shard wildcard, built-in type": {
+			cluster:       genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
+			shard:         "*",
+			key:           "/registry/group/resource/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/",
+			expectedShard: "amber",
+		},
+		"cluster wildcard, built-in type": {
+			cluster:       genericapirequest.Cluster{Wildcard: true},
+			shard:         "amber",
+			key:           "/registry/group/resource/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/amber/",
+			expectedShard: "amber",
+		},
+		"both wildcard, built-in type": {
+			cluster:       genericapirequest.Cluster{Wildcard: true},
+			shard:         "*",
+			key:           "/registry/group/resource/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/",
+			expectedShard: "amber",
+		},
+
+		// CRD types.
+
+		"not wildcard, CRD": {
+			crdRequest:    true,
+			cluster:       genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
+			shard:         "amber",
+			key:           "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/identity/amber/root:org:ws/",
+			expectedShard: "amber",
+		},
+
+		"both wildcard, CRD": {
+			crdRequest:    true,
+			cluster:       genericapirequest.Cluster{Wildcard: true},
+			shard:         "*",
+			key:           "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/identity/",
+			expectedShard: "amber",
+		},
+		"both wildcard, CRD, partial metadata request": {
+			crdRequest:    true,
+			cluster:       genericapirequest.Cluster{Wildcard: true, PartialMetadataRequest: true},
+			shard:         "*",
+			key:           "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/",
+			expectedShard: "amber",
+		},
+
+		"only cluster wildcard, CRD": {
+			cluster:       genericapirequest.Cluster{Wildcard: true},
+			shard:         "amber",
+			crdRequest:    true,
+			key:           "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/",
+			expectedShard: "amber",
+		},
+		"only cluster wildcard, CRD, partial metadata request": {
+			cluster:       genericapirequest.Cluster{Wildcard: true, PartialMetadataRequest: true},
+			shard:         "amber",
+			crdRequest:    true,
+			key:           "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/",
+			expectedShard: "amber",
+		},
+
+		"only shard wildcard, CRD": {
+			cluster:       genericapirequest.Cluster{Name: logicalcluster.Name("root:org:ws")},
+			shard:         "*",
+			crdRequest:    true,
+			key:           "/registry/group/resource/identity/amber/root:org:ws/somename",
+			keyPrefix:     "/registry/group/resource/identity/",
+			expectedShard: "amber",
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			shardName := adjustShardNameIfWildcard(tc.shard, &tc.cluster, tc.crdRequest, tc.keyPrefix, tc.key)
+			if tc.expectedShard != shardName {
+				t.Errorf("expected: %q, actual %q", tc.expectedShard, shardName)
+			}
+		})
+	}
+}
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go
--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go
@@ -734,7 +734,7 @@ func (wc *watchChan) prepareObjs(e *event) (curObj runtime.Object, oldObj runtim
 
 		// kcp: apply clusterName to the decoded object, as the name is not persisted in storage.
 		clusterName := adjustClusterNameIfWildcard(wc.shard, wc.cluster, wc.crdRequest, wc.key, e.key)
-		shardName := adjustShardNameIfWildcard(wc.shard, wc.key, e.key)
+		shardName := adjustShardNameIfWildcard(wc.shard, wc.cluster, wc.crdRequest, wc.key, e.key)
 		annotateDecodedObjectWith(curObj, clusterName, shardName)
 	}
 	// We need to decode prevValue, only if this is deletion event or
@@ -756,7 +756,7 @@ func (wc *watchChan) prepareObjs(e *event) (curObj runtime.Object, oldObj runtim
 
 		// kcp: apply clusterName to the decoded object, as the name is not persisted in storage.
 		clusterName := adjustClusterNameIfWildcard(wc.shard, wc.cluster, wc.crdRequest, wc.key, e.key)
-		shardName := adjustShardNameIfWildcard(wc.shard, wc.key, e.key)
+		shardName := adjustShardNameIfWildcard(wc.shard, wc.cluster, wc.crdRequest, wc.key, e.key)
 		annotateDecodedObjectWith(oldObj, clusterName, shardName)
 	}
 	return curObj, oldObj, nil
